### PR TITLE
Freeze trailing row support

### PR DESCRIPTION
**Pull request by jassmith**
```[tasklist]
### Needed fixes
- [x] Fix rendering of highlight rings that span the freeze rows
- [x] The visibile region call needs to properly expose the frozen rows
- [x] Strict mode needs to be updated to allow `getCellContent` of these rows
- [x] Check ButtonCell animation works
- [x] Write a proper story
- [x] Fix renderStateProvider
- [x] When scrolled to bottom the grid line is double heavy (lo-dpi only) !needs confirmation of fix!
- [ ] Accessibility might need update
- [x] Scroll to does not properly account for frozen rows.
- [x] Issue with drawing of borders (see comment below)
- [x] Column selection doesn't include frozen rows
```

### CODE REVIEW

File: packages/core/src/internal/data-grid/data-grid-lib.ts [hunk 1]
```diff
@@ -155,6 +155,18 @@ export function getStickyWidth(
...
 }
 
+export function getFreezeTrailingHeight(
+    rows: number,
+    freezeTrailingRows: number,
+    getRowHeight: number | ((row: number) => number)
+): number {
+    let result = 0;
+    for (let i = rows - freezeTrailingRows; i < rows; i++) {
+        result += typeof getRowHeight === "number" ? getRowHeight : getRowHeight(i);
```
Comment (jassmith) [inline packages/core/src/internal/data-grid/data-grid-lib.ts:165]: This is inefficient for large numbers of freeze regions, simple multiplication can work anytime the rowHeight is fixed which is the vast majority of cases.

File: packages/core/src/internal/data-grid/data-grid-lib.ts [hunk 2]
```diff
@@ -224,18 +236,23 @@ export function getRowIndexForY(
...
 ): number | undefined {
     const totalHeaderHeight = headerHeight + groupHeaderHeight;
     if (hasGroups && targetY <= groupHeaderHeight) return -2;
     if (targetY <= totalHeaderHeight) return -1;
 
-    const lastRowHeight = typeof rowHeight === "number" ? rowHeight : rowHeight(rows - 1);
-    if (lastRowSticky && targetY > height - lastRowHeight) {
-        return rows - 1;
+    let y = height;
+    for (let fr = 0; fr < freezeTrailingRows; fr++) {
```
Comment (jassmith) [inline packages/core/src/internal/data-grid/data-grid-lib.ts:246]: Exactly how this loop works might need some comments

File: packages/core/src/internal/data-grid/data-grid-render.tsx [hunk 3]
```diff
@@ -470,22 +468,33 @@ function drawGridLines(
         }
     }
 
-    const stickyHeight = getRowHeight(rows - 1);
-    const stickyRowY = height - stickyHeight + 0.5;
-    const lastRowSticky = trailingRowType === "sticky";
-    if (lastRowSticky) {
-        toDraw.push({ x1: minX, y1: stickyRowY, x2: maxX, y2: stickyRowY, color: hColor });
+    // here we need to push all the relevant lines for the sticky row
+    // const stickyHeight = getRowHeight(rows - 1);
```
Comment (jassmith) [inline packages/core/src/internal/data-grid/data-grid-render.tsx:472]: delete me

File: packages/core/src/internal/data-grid/data-grid-render.tsx [hunk 4]
```diff
@@ -470,22 +468,33 @@ function drawGridLines(
...
-        const target = lastRowSticky ? height - stickyHeight : height;
+        const target = freezeY;
         while (y + translateY <= target) {
             const ty = y + translateY;
             // This shouldn't be needed it seems like... yet it is. We're not sure why.
-            if (ty >= minY && ty <= maxY - 1 && (!lastRowSticky || row !== rows - 1 || Math.abs(ty - stickyRowY) > 1)) {
+            if (
+                ty >= minY &&
+                ty <= maxY - 1 &&
+                (freezeTrailingRows === 0 || row !== rows - 1 || Math.abs(ty - freezeY) > 1)
```
Comment (jassmith) [inline packages/core/src/internal/data-grid/data-grid-render.tsx:496]: It is unclear at this time if this is working correctly.

File: packages/core/src/internal/data-grid/data-grid-render.tsx [hunk 5]
```diff
@@ -1181,7 +1191,8 @@ function drawCells(
     ctx.font = font;
     const deprepArg = { ctx };
     const cellIndex: [number, number] = [0, 0];
-    const stickyRowHeight = trailingRowType === "sticky" ? getRowHeight(rows - 1) : 0;
+    const stickyRowHeight =
```
Comment (jassmith) [inline packages/core/src/internal/data-grid/data-grid-render.tsx:1194]: stickyRowHeight is a bad name now

File: packages/core/src/internal/data-grid/data-grid-render.tsx [hunk 6]
```diff
@@ -2726,7 +2657,10 @@ export function drawGrid(arg: DrawGridArg, lastArg: DrawGridArg | undefined) {
             width: effectiveCols.length,
             height: lastRowDrawn - cellYOffset,
         },
-        freezeColumns
+        freezeColumns,
+        Array.from({ length: freezeTrailingRows })
```
Comment (jassmith) [inline packages/core/src/internal/data-grid/data-grid-render.tsx:2661]: This is probably inefficient

File: packages/core/src/data-editor/data-editor.tsx [hunk 7]
```diff
@@ -2347,6 +2367,37 @@ const DataEditorImpl: React.ForwardRefRenderFunction<DataEditorRef, DataEditorPr
...
+                freezeColumns === 0
+                    ? undefined
+                    : {
+                          x: 0,
+                          y: region.y,
+                          width: freezeColumns,
+                          height: region.height,
+                      };
+
+            const freezeRegions: Rectangle[] = [];
```
Comment (LukasMasuch) [inline packages/core/src/data-editor/data-editor.tsx:2381]: Does this theoretically support freezing any combination of cells?

Comment (jassmith) [reply packages/core/src/data-editor/data-editor.tsx:2381]: its setup so that we can make that happen

**Comment by LukasMasuch** [comment]
From discord conversation: there seems to be a problem with rendering of borders, but not sure if this is specific to this branch:

https://github.com/glideapps/glide-data-grid/assets/2852129/e224a151-0415-4f4d-a514-dda4a2f87695

**Comment by jassmith** [comment]
Fixed :)

**Comment by LukasMasuch** [comment]
Another issue seems to be with column selection not including frozen rows:

https://github.com/glideapps/glide-data-grid/assets/2852129/a29cc690-8570-4c17-aa5a-5ca4cb02f9b9

**Comment by jassmith** [comment]
Good catch!

**Review by LukasMasuch — APPROVED** [review]
Code looks good as far as I understood it 👍  

Two additional thoughts on this: 

1) maybe it would also benefit from a shadow like frozen columns and header row?
2) I think some of the uses for frozen rows will require to exclude frozen rows from some of the operations similar to the append row (cell highlighting, rowMarker, fill handle, ..). E.g. if it just shows some column statistics, it won't be really useful to select it together with the actual data above. This is also similar to [how aggrid handles pinned rows](https://www.ag-grid.com/javascript-data-grid/row-pinning/#non-supported-items). But that could be another config option to exclude those rows. I think this might be very similar to how grouped rows also need to be excluded from some operations as discussed in: https://github.com/glideapps/glide-data-grid/issues/823

**Pull request merged**